conversion/conversion.go

```diff
@@ -460,26 +467,52 @@ func UpdateDDLForeignKeys(project, instance, dbName string, conv *internal.Conv,
 	}
 	msg := fmt.Sprintf("Updating schema of database %s in instance %s with foreign key constraints ...", dbName, instance)
 	p := internal.NewProgress(int64(len(fkStmts)), msg, internal.Verbose())
-	for i, fkStmt := range fkStmts {
-		// TODO: Improve performance of the foreign key constraint updates.
-		// For example, issue all of the update ops first before waiting for them to complete
-		// so that that can execute in parallel. We could also print out ids of the
-		// long-running operations.
-		op, err := adminClient.UpdateDatabaseDdl(ctx, &adminpb.UpdateDatabaseDdlRequest{
-			Database:   fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, dbName),
-			Statements: []string{fkStmt},
-		})
-		if err != nil {
-			fmt.Printf("Can't add foreign key with statement %s: %s\n", fkStmt, err)
-			conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
-			continue
-		}
-		if err := op.Wait(ctx); err != nil {
-			fmt.Printf("Can't add foreign key with statement %s: %s\n", fkStmt, err)
-			conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
-		}
-		internal.VerbosePrintln("Updated schema with statement: " + fkStmt)
-		p.MaybeReport(int64(i + 1))
+
+	workers := make(chan int, MaxWorkers)
+	for i := 1; i <= MaxWorkers; i++ {
+		workers <- i
+	}
+	var progressMutex sync.Mutex
+	progress := int64(0)
+
+	// We dispatch parallel foreign key create requests to ensure the backfill runs in parallel to reduce overall time.
+	// This cuts down the time taken to a third (approx) compared to Serial and Batched creation. We also do not want to create
+	// too many requests and get throttled due to network or hitting catalog memory limits.
+	// Ensure atmost `MaxWorkers` go routines run in parallel that each update the ddl with one foreign key statement.
+	for _, fkStmt := range fkStmts {
+		workerId := <-workers
+		go func(fkStmt string, workerId int) {
+			defer func() {
+				workers <- workerId
```

Comment:
Shouldn't this be the last thing the worker does?

Suggest moving this line to the end of this block i.e. after line 492.

That way you are faithful to your comment "ensure at most maxworker go routines run in parallel".

Note that what you have is perfectly correct, since what we really want to protect is the number of outstanding FK create requests and the current code does that. I'm more thinking about keeping the code simple, following expected patterns, and having code be in sync with comments.

Reply:
Agree on that